### DISCUSSION TITLE
Benchmarking: Overestimated surface displacement?

### DISCUSSION THREAD

**Discussion by tshreve** [idaholab/moose#18898] [Q&A Modules: Solid mechanics] [answered]
I am attempting to benchmark Moose's Tensor Mechanics module for a spheroidal pressurized cavity against an analytical solution developed for volcanic source modelling (https://pubs.usgs.gov/tm/13/b1/pdf/tm13-b1.pdf). Assuming linear elasticity, I would expect surface displacements to scale linearly with pressure change, however I found the surface displacements are overestimated when imposing a high pressure change (~1 GPa) on the cavity boundary. Figure 1 shows surface displacements when the boundary pressure change is ~1 GPa, and Figure 2 shows surface displacements when it is ~10 MPa (surface displacements are then scaled by 100). I was wondering if you have any insights as to where this discrepancy comes from.

Thanks in advance for your help!

**Figure 1:**
![Fig1_highDP](https://user-images.githubusercontent.com/25572128/134399280-dbccddf3-20bb-4f07-b4b2-0a29fb540693.png)

**Figure 2:**
![Fig2_scaledDP](https://user-images.githubusercontent.com/25572128/134399283-6b5b9250-6aa8-4d48-a971-adef1bb32f1b.png)

**Comment by GiudGiud** [comment]
Hello
Just a small question and I ll tag the experts after, how refined is your mesh? Did you try using the uniform_refine parameter in [Mesh]?

**Reply by tshreve** [reply]
The mesh is refined to ~80 m at the surface in a 2 km radius around the cavity, then gradually reduces to ~6 km at a distance of 30 km away from the source. The same mesh is used for both calculations and was created using gmsh. I have not tried using uniform_refine since I am already reaching a high number of DOFs (`Num DOFs:                274857`, `Num Local DOFs: 37722` if running on 8 cores).
![meshSpheroid](https://user-images.githubusercontent.com/25572128/134402001-50875a10-cc21-4021-993b-9d178751655b.png)

**Reply by GiudGiud** [reply]
it would be great to check that a single level of refinement does not change the answer much

**Comment by GiudGiud** [comment]
@bwspenc @dschwen @jiangwen84

**Comment by jiangwen84** [comment, marked as answer]
Couple of things:

1. please share your input file.
2. is the analytical solution derived based on small strain or finite strain assumption? I expect the the deformation will be larger for higher pressure and it might change the assumption.  Could you provide the analytical solution here?
3. Is the domain large enough to eliminate the BC effect? I guess the analytical solution assumes infinite domain.

**Reply by tshreve** [reply]
1. Input file for high pressure example is attached.
2. It is based on small strain assumption and is described on page 20 (Prolate Spheroid Source) of this document (https://pubs.usgs.gov/tm/13/b1/pdf/tm13-b1.pdf). Let me know if you need more info.
3. The domain is 100 km while the wavelength of the signal is <15 km, so we expect the BC effect to be mitigated.

[Yang_ElastA3d_nograv_a1000_b500_t90_p0_vertical_highP.txt](https://github.com/idaholab/moose/files/7213114/Yang_ElastA3d_nograv_a1000_b500_t90_p0_vertical_highP.txt)

**Reply by ngrilli** [reply]
Dear @tshreve 

As others say, try a bigger domain, especially in depth along z,
for bigger pressures the effect of your "signal wavelength" may be larger and you may need a larger domain.

Also, try hexahedral elements, tetrahedra elements are usually not very accurate for problems involving bending.

Also, check the boundary conditions assumption in the analytical model,
I think the pressure BC in MOOSE takes into account of the deformed surface,
meaning the total force on the surface of one boundary element is pressure * deformed area
 
while in the analytical model the boundary condition may not consider the deformation,
it may be applied on the undeformed domain,
otherwise, I am not sure if an analytical solution would be possible. 

Nicolò

**Reply by jiangwen84** [reply]
I agree with @ngrilli . 

1. Add `use_displaced_mesh = false` in the Pressure block.  
2. do you have other 2D example? Maybe you can use them to see the boundary effect. Currently you use symmetric BC, maybe try use a minimum BC that just restricts the rigid body motion. 
3. What did you plot in those figures? displacement along the center line? I suggest you use `LineValueSampler` to get them directly from MOOSE.

**Reply by tshreve** [reply]
Thanks for your quick responses!

1. After running with `use_displaced_mesh = false` in the Pressure block, I obtain the following: `Warning: unused parameter 'BCs/Pressure/use_displaced_mesh'.` Do you know why this parameter is not be taken into consideration? We also noticed that in the [Pressure BC documentation](https://mooseframework.inl.gov/source/bcs/Pressure.html) the pressure is always applied to the displaced mesh, which we suspect would explain the discrepancy in the displacements.

To answer your other questions:

2. I tried with a 2D example and with a large enough domain (100 km) did not see any BC effects.
3. I plot the surface displacements with python after testing various interpolation schemes. I will try `LineValueSampler` for comparison.

**Reply by jiangwen84** [reply]
@tshreve The `use_displaced_mesh` option is not implemented in `Pressure Action`. You can use

```
  [pressure_x]
    type = Pressure
    boundary = chamber
    variable = disp_x
    component = 0
    factor = 0.96e9
    use_displaced_mesh = false
  []

  [pressure_y]
    type = Pressure
    boundary = chamber
    variable = disp_y
    component = 1
    factor = 0.96e9
    use_displaced_mesh = false
  []

  [pressure_z]
    type = Pressure
    boundary = chamber
    variable = disp_z
    component = 2
    factor = 0.96e9
    use_displaced_mesh = false
  []

```

**Reply by tshreve** [reply]
Thanks, I'll give that a try!

**Reply by tshreve** [reply]
`use_displaced_mesh = false` resolved the discrepancy. Thanks, we appreciate your help!

**Comment by hugary1995** [comment]
`use_displaced_mesh` probably shouldn't be suppressed in the Pressure action.

**Reply by jiangwen84** [reply]
It is not suppressed, just not implemented as an option in Pressure action. I agree we should add it, see #18979

**Reply by dschwen** [reply]
Exposing it would just be another chance for users to shoot themselves in the foot, wouldn't it? Physically it makes no sense IMO to apply the forces computed on the undisplaced mesh to the displaced mesh for a pressure BC.

**Reply by jiangwen84** [reply]
If we set 'use_displaced_mesh = false`, we will calculate the forces on the undisplayed mesh and apply it on the **undisplayed** mesh too. Isn't it? 

I guess another way to resolve @tshreve 's issue is NOT to use displaced mesh by setting 'use_displaced_mesh = false` in the Mesh block.

**Reply by hugary1995** [reply]
> Exposing it would just be another chance for users to shoot themselves in the foot, wouldn't it? Physically it makes no sense IMO to apply the forces computed on the undisplaced mesh to the displaced mesh for a pressure BC.

I think that depends. The root cause of the issue that @tshreve 's seeing is that she is assuming small deformation, while although the deformation is in some sense small, it is _large enough_ to make a non-negligible mismatch with the analytical solution. Ideally, if the deformation was small enough, `use_displaced_mesh` should make negligible difference.

Now, there are several ways to solve this issue, as @jiangwen84 suggested:
1. Do not use the Pressure action, but setup pressure BCs manually with `use_displaced_mesh = false`,
2. Let the displaced mesh fall back to the undisplaced mesh, by setting `use_displaced_mesh = false` in the Mesh block,
3. Reduce the magnitude of the pressure until the mismatch between the numerical solution and the analytical solution is negligible.

Then, as a physics module, we need to make a choice:
a. If we want to support methods 1 and 2, then we better allow actions to set those parameters;
b. If we suggest method 3, we are basically saying this is user error -- the user is using linear elasticity but the problem is violating the small strain assumptions.

I see arguments for both choices. I guess if we make choice b, it would be better to issue runtime warnings like "You are using a small strain formulation,but the deformation in your problem is outside the small strain regime, be aware of numerical errors...".

**Reply by dschwen** [reply]
Ok, makes sense. Wen, I think you're right. The force should be applies to the undisplaced mesh. I guess that means the force is still applied normal to the surface, but the local magnitude of the force is not quite correct because the undeformed  element area rather than the deformed element are is used, right?
